File: Assets/Game/Scripts/Scene/Terrain/Terrains.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Terrains
{
    protected enum Axial
    {
        X,
        Y,
        Z
    }

    protected int m_FloorCount;
    protected int m_Width;
    protected int m_Height;
    protected int m_FloorPieceCount;

    protected List<TerrainPiece> m_TerrainPieces;
    public Terrains()
    {
        m_FloorPieceCount = m_Width * m_Height;
    }

    private int SpacePositionToId(Vector3 spacePos)
    {
        return (int) (spacePos.x + m_Width * spacePos.z + m_FloorPieceCount * spacePos.y);
    }

    private Vector3 FloorPositionCorrection(Vector3 spacePosition)
    {
        Vector3 newPos = spacePosition;
        newPos = SpaceDirectionCorrection(Axial.X, newPos);
        newPos = SpaceDirectionCorrection(Axial.Z, newPos);
        return newPos;
    }

    //-
    private Vector3 SpaceDirectionCorrection(Axial axial, Vector3 spacePosition)
    {
        float value = axial == Axial.X ? spacePosition.x : (axial == Axial.Y ? spacePosition.y : spacePosition.z);
        float round = axial == Axial.X ? m_Width : (axial == Axial.Y ? m_Height : m_FloorCount);

        bool update = false;

        if (value < 0)
        {
            value = 0;
            update = true;
        }
        else if (value >= round)
        {
            value = round;
            update = true;
        }

        if (update)
        {
            Vector3 newSpacePos = spacePosition;
            UpdateSpaceValue(axial, newSpacePos, value);

            int pieceId = SpacePositionToId(newSpacePos);
            TerrainPiece curPiece = m_TerrainPieces[pieceId];
            if (curPiece.GetSpace() == TerrainPiece.Space.Loop)
            {
                UpdateSpaceValue(axial, newSpacePos, value);
            }
        }

        return spacePosition;
    }

    private Vector3 UpdateSpaceValue(Axial axial, Vector3 position,float value)
    {
        if (axial == Axial.X)
        {
            position.x = value;
        }
        else if (axial == Axial.Y)
        {
            position.y = value;
        }
        else
        {
            position.z = value;
        }

        return position;
    }

    public Vector3 GetNextSpacePosition(Vector3 curSpacePos, Vector3 direction)
    {
        Vector3 nextSpacePos = curSpacePos;
        direction.y = 0;
        nextSpacePos += direction;
        nextSpacePos = FloorPositionCorrection(nextSpacePos);

        bool loop = false;
        Vector3 lastPos = nextSpacePos;
        do
        {
            loop = false;

            int pieceId = SpacePositionToId(nextSpacePos);
            TerrainPiece curPiece = m_TerrainPieces[pieceId];
            TerrainPiece.Direction pieceDirection = curPiece.GetDirection();

            if (pieceDirection == TerrainPiece.Direction.Down)
            {
                nextSpacePos += Vector3.up;
                loop = true;
            }
            else if (pieceDirection == TerrainPiece.Direction.Up)
            {
                nextSpacePos += Vector3.down;
                loop = true;
            }
            else if (pieceDirection == TerrainPiece.Direction.Not)
            {
                nextSpacePos = lastPos;
            }

            nextSpacePos = SpaceDirectionCorrection(Axial.Y, nextSpacePos);
            lastPos = nextSpacePos;

        } while (loop);


        return nextSpacePos;
    }
}

File: Assets/Game/Lua/General/Terrain/Terrains.lua
Terrains = class()

require "General/SceneObject/TerrainPiece/TerrainPiece"

function Terrains:ctor()
    self.pTerrainPieces = {}
end

function Terrains:Create(terrainData)
    self.pTerrainData = terrainData
    self.pWidth = terrainData.Width
    self.pHeight = terrainData.Height
    self.pFloorCount = terrainData.FloorCount
    self.pFloorPieceCount = self.pWidth * self.pHeight

    local x = 0
    local z = 1
    local pieceCount = 1

    local pieceSize = Vector3.New(terrainData.Building.PieceSize.Width, 0, terrainData.Building.PieceSize.Height) 
    local pieceSizeRaius = pieceSize * 0.5

    for floorCount = 0, self.pFloorCount - 1 do 
        local initPos = Vector3.New(terrainData.Building.SideGap.Width, -floorCount * terrainData.Building.FloorHeight, terrainData.Building.SideGap.Height ) 
        
        for floorPieceCount = 1, self.pFloorPieceCount do

            local pieceData = terrainData.Piece[pieceCount]

            local curPos = initPos + Vector3.New(pieceSizeRaius.x + pieceSize.x * x, 0, pieceSizeRaius.z )
            local piece = self:CreateTerrainPiece(pieceData.Id, curPos, pieceCount, Vector3.New(x, floorCount, z), pieceData.Direction, pieceData.Spcae )
            self.pTerrainPieces[pieceCount] = piece

            x = x + 1
            if x % self.pWidth == 0 then
                x = 0
                z = z + 1
                initPos.x = terrainData.Building.SideGap.Width
            else
                initPos.x = initPos.x + terrainData.Building.PieceGap.Width
            end

            if z % self.pHeight == 0 then
                z = 1
                initPos.z = initPos.z + pieceSize.z * z + terrainData.Building.PieceGap.Height 
            end

            pieceCount = pieceCount + 1
        end
    end
end


function Terrains:CreateTerrainPiece(id, position, index, spacePosition, direction, spcae)
    local piece = TerrainPiece:new(id, position, index, spacePosition, direction, spcae)
    return piece
end

--得到下一个空间位置
function Terrains:GetNextSpacePosition(curSpacePos, direction)
    local nextSpacePos = curSpacePos

    direction.y = 0
    nextSpacePos = nextSpacePos:Add(direction)
    nextSpacePos = self:FloorPositionCorrection(nextSpacePos)

    local loop = true
    local lastPos = nextSpacePos

    while( loop == true )
    do
        loop = false
        
        local pieceId = self:SpacePositionToId(nextSpacePos)
        local curPiece = self.m_TerrainPieces[pieceId]
        local pieceDirection = curPiece.GetDirection()

        if pieceDirection == GameDefine.TerrainPiece.Direction.Down then
            nextSpacePos = nextSpacePos:Add(Vector3.up)
            loop = true
        elseif pieceDirection == GameDefine.TerrainPiece.Direction.Up then
            nextSpacePos = nextSpacePos:Add(Vector3.down)
            loop = true
        elseif pieceDirection == GameDefine.TerrainPiece.Direction.Not then
            nextSpacePos = lastPos
        end

        nextSpacePos = self:SpaceDirectionCorrection(GameDefine.Terrain.Axial.Y, nextSpacePos)
        lastPos = nextSpacePos
    end

    return nextSpacePos
end

function Terrains:SpacePositionToId(spacePos)
    return spacePos.x + self.pWidth * spacePos.z + self.pFloorPieceCount * spacePos.y 
end

function Terrains:FloorPositionCorrection(spacePosition)
    local newPos = spacePosition
    newPos = self:SpaceDirectionCorrection(GameDefine.Terrain.Axial.X, newPos)
    newPos = self:SpaceDirectionCorrection(GameDefine.Terrain.Axial.Z, newPos)
    return newPos
end

function Terrains:SpaceDirectionCorrection(axial, spacePosition)
    local value = spacePosition.x 
    local round = self.pWidth
    if axial == GameDefine.Terrain.Axial.Z then
        value = spacePosition.z 
        round = self.pHeight
    elseif axial == GameDefine.Terrain.Axial.Y then
        value = spacePosition.y 
        round = self.pFloorCount
    end

    local update = false
    if value < 0 then
        value = 0
        update = true
    elseif value >= round then
        value = round
        update = true
    end

    if update then
        local newSpacePos = spacePosition
        self:UpdateSpaceValue(axial, newSpacePos, value)

        local pieceId = self:SpacePositionToId(newSpacePos)
        local curPiece = self.pTerrainPieces[pieceId]
        if curPiece.GetSpace() == GameDefine.TerrainPiece.Space.Loop then
            self:UpdateSpaceValue(axial, newSpacePos, value)
        end
    end

    return spacePosition
end

function Terrains:UpdateSpaceValue(axial, position, value)
    if axial == GameDefine.Terrain.Axial.X then
        position.x = value
    elseif axial == GameDefine.Terrain.Axial.Y then
        position.y = value
    else
        position.z = value
    end

    return position
end

